### PR TITLE
Add menu bar mode for macOS

### LINKED ISSUE #206: Add menu bar icon on macOS
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/caprine/issues/206)

It'd be cool to show the number of unread notifications and add the ability to remove the app from dock. Something like this: 
dock_icon_preview
menu_messenger_icon

The number may be too small in my example, but I think the main idea is there.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $2.00)

- [rororofff rororofff](https://issuehunt.io/u/rororofff) ($2.00)
### Submitted pull Requests
- [#804 Add menu bar mode for macOS](https://issuehunt.io/r/sindresorhus/caprine/pull/804)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/caprine/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by whitecrownclown**
@sindresorhus @CvX This PR continues the work from #374

If you could please have a look, I've fixed most/all of the issues @sindresorhus addressed.

Thanks!

Fixes #206

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#206: Add menu bar icon on macOS](https://issuehunt.io/repos/42574339/issues/206)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by CvX** [comment]
Not a full review yet, just some thoughts after trying this out for the first time:

1. The menubar icon's shape should be based on the current app icon, i.e. not as wide. Also (this is quite subjective) I'd make the outline thicker, add the Messenger's "zig-zag" to both icon states and would make the unread one blue-ish (like e.g. Tweetbot: Tweetbot menu bar icon)
2. I'd replace any mentions of "darwin" with just "macos"
3. Would move as much of new code (e.g. `setupMenuBarMode`) to either `tray.ts` or a new file to keep it encapsulated (for an example, see `emoji.ts`).
4. "Disable Menu Bar Mode" closes the app

Other than that, looks cool! 🙂

**Comment by whitecrownclown** [comment]
@CvX Hmm, I was almost sure I took care of 4). Thanks for the suggestions and obviously, your time, I'll update the PR.

File: source/tray.ts [hunk 1]
```diff
@@ -59,10 +59,13 @@ export default {
 	},
 
 	destroy: () => {
-		if (tray) {
-			tray.destroy();
-			tray = null;
-		}
+		// Workaround for electron#14036
```
Comment (sindresorhus) [inline source/tray.ts]: Can you use a full link to the issue, so it's more easily clickable?

**Comment by sindresorhus** [comment]
If I activate menu bar mode, there's no way to access the menu items. I think it should still show the menu when the window is focused.

**Comment by sindresorhus** [comment]
> The menubar icon's shape should be based on the current app icon, i.e. not as wide. 

👍 

> Also (this is quite subjective) I'd make the outline thicker

👍 

> add the Messenger's "zig-zag" to both icon states and would make the unread one blue-ish (like e.g. Tweetbot: Tweetbot menu bar icon)

👎 The "zig-zag" is the indicator of a new message. It should stay. Menu bar icons should not be colored.

---

The menu bar icon is also slightly to large and not fully vertically centered.

Screen Shot 2019-03-10 at 23 17 13

**Comment by whitecrownclown** [comment]
Thanks, i’ll update it this week!

**Comment by whitecrownclown** [comment]
> If I activate menu bar mode, there's no way to access the menu items. I think it should still show the menu when the window is focused.

Apparently there's no way to do this at the moment. When calling `app.dock.hide()`, the menu items are also hidden. [issue](https://github.com/electron/electron/issues/592)

**Comment by whitecrownclown** [comment]
@sindresorhus 	I've updated the icons. This is how they look now.

`read`
read

`unread`
unread

**Comment by sindresorhus** [comment]
> Apparently there's no way to do this at the moment. When calling app.dock.hide(), the menu items are also hidden.

Hmm, that's too bad. It would mean there's no way to access the menus, and also no keyboard shortcuts would work as they are registered in the menu...

**Comment by sindresorhus** [comment]
The menu bar icon is still too large and it should be updated to the style of the app icon, meaning softer lightening symbol (just copy-paste it from the icon), softer arrow under bottom-left of the circle, and the circle shape is different.

**Comment by whitecrownclown** [comment]
@sindresorhus I've updated the icons, size is 16px (previously 18px). I have the vector assets now so if we need changes, it'll be easier. The icon shape should be basically the same as `caprine` icon. Couldn't find a vector for it so i traced the png in inkscape instead. Let me know what you think, thanks!

`read`
read

`unread`
unread

**Comment by sindresorhus** [comment]
> Couldn't find a vector for it so i traced the png in inkscape instead.

It's right here: https://github.com/sindresorhus/caprine/blob/master/media/Icon.svg

**Comment by sindresorhus** [comment]
The icon positions are not the same:

foo

**Comment by whitecrownclown** [comment]
> > Couldn't find a vector for it so i traced the png in inkscape instead.
> 
> It's right here: [/media/Icon.svg@`master`](https://github.com/sindresorhus/caprine/blob/master/media/Icon.svg)

@sindresorhus 	Thanks, I've updated it based on the same app icon `media/Icon.svg`.

menubar

**Comment by sindresorhus** [comment]
We still need to find a way to resolve: https://github.com/sindresorhus/caprine/pull/804#issuecomment-473781977

How about if "menu bar mode" doesn't imply hiding the Dock icon? When the user enables menu bar mode, we enable a menu item that allows hiding the Dock icon, and when the user selects that, we show a dialog letting them know that keyboard shortcuts will not work. We also let the user access the normal menu from the menu bar icon menu:

```
Disable Menu Bar Mode
--------------
Menu >
--------------
Quit
```

// @CvX Thoughts?

**Comment by sindresorhus** [comment]
> Thanks, I've updated it based on the same app icon media/Icon.svg.

👍 Looks good.

**Comment by CvX** [comment]
Yeah, splitting that into two separate options is common in macOS apps: "Show Menu Bar icon" and "Show Dock icon"

Here's how it appears in Fantastical preferences:
Fantastical options

I'm for it! Adding all app menus to the menu bar icon's context menu is also a great idea.

But it would be nice to make keyboard shortcuts work also without the menu bar.

**Comment by sindresorhus** [comment]
> Yeah, splitting that into two separate options is common in macOS apps: "Show Menu Bar icon" and "Show Dock icon"

👍 

> But it would be nice to make keyboard shortcuts work also without the menu bar.

I'm open to ideas. I don't see how we can do that though... Maybe a `Electron.MenuItem` wrapper that also register accelerators using https://www.npmjs.com/package/electron-localshortcut when in menubar mode. Although, that's ugly. Hoping for a better solution than that. We can tackle that later on though.

**Comment by sindresorhus** [comment]
@whitecrownclown This is so close to getting merged. Interested in finishing it up?

**Comment by whitecrownclown** [comment]
Yes, will work on it.

**Comment by whitecrownclown** [comment]
@sindresorhus I've added the menu into the tray when enabling `MenuBarMode` and the shortcuts seem to work just fine. Please have a look.

**Comment by CvX** [comment]
I've tried it out and it's looking great! 😃🎉

We're still going through with splitting the "Menu bar mode" into two separate options ("Show Menu Bar icon" and "Show Dock icon"), right?

**Comment by vanshg** [comment]
@whitecrownclown Was just about to pick back up where I left off on https://github.com/sindresorhus/caprine/pull/374, when I saw this. Thank you for continuing this work!

**Comment by whitecrownclown** [comment]
@vanshg Hey, thanks for starting #374 

I've separated the two options. When toggling `Show Menu Bar Icon` it will enable `Show Dock Icon` and it works, but I have one issue : 

If user has `Show Menu Bar Icon` true and `Show Dock Icon` false, when restarting `Caprine`, the app will start with the dock icon and then it would need to be hidden using : `app.dock.hide()` and it doesn't look great. I couldn't find a way around except this https://github.com/electron-userland/electron-builder/issues/1456#issuecomment-293518692 but i don't think it applies to this use case.

@CvX @sindresorhus Thoughts? Also have a look at the changes. Thanks!

**Comment by sindresorhus** [comment]
If we set https://github.com/electron-userland/electron-builder/issues/1456#issuecomment-293518692, can we then show the Dock icon? If so, we could just do the reverse of what we're doing.

File: readme.md [hunk 2]
```diff
@@ -19,6 +19,7 @@ Caprine is an unofficial and privacy-focused Facebook Messenger app with many us
 - [Vibrant theme](#vibrancy)\*
 - [Privacy-focused](#privacy)
 - [Keyboard shortcuts](#keyboard-shortcuts)
+- [Menu Bar mode](#menu-bar-mode)
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
- [Menu Bar mode](#menu-bar-mode)\*
```

File: source/index.ts [hunk 3]
```diff
@@ -86,13 +87,17 @@ app.on('second-instance', () => {
...
+}
+
 function updateBadge(conversations: Conversation[]): void {
 	// Ignore `Sindre messaged you` blinking
 	if (!Array.isArray(conversations)) {
 		return;
 	}
 
-	const messageCount = conversations.filter(({unread}) => unread).length;
+	const messageCount: number = getMessageCount(conversations);
```
Comment (sindresorhus) [inline source/index.ts]:
```suggestion
	const messageCount = getMessageCount(conversations);
```

File: source/index.ts [hunk 4]
```diff
@@ -31,6 +31,7 @@ import {sendAction, sendBackgroundAction} from './util';
 import {process as processEmojiUrl} from './emoji';
 import ensureOnline from './ensure-online';
 import './touch-bar'; // eslint-disable-line import/no-unassigned-import
+import {setupMenuBarMode} from './menu-bar-mode';
```
Comment (sindresorhus) [inline source/index.ts]:
```suggestion
import {setUpMenuBarMode} from './menu-bar-mode';
```

File: source/menu-bar-mode.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import {app, globalShortcut, BrowserWindow, Menu} from 'electron';
+import {is} from 'electron-util';
+
+import config from './config';
+import tray from './tray';
+
+const menuBarShortcut = 'CommandOrControl+Shift+Y';
```
Comment (sindresorhus) [inline source/menu-bar-mode.ts]: This is only for macOS, so we don't need `Control`.

Comment (sindresorhus) [reply source/menu-bar-mode.ts]: This needs to be documented in the keyboard and menu bar mode section in the readme.

File: source/tray.ts [hunk 6]
```diff
@@ -1,44 +1,84 @@
 import * as path from 'path';
 import {app, Menu, Tray, BrowserWindow} from 'electron';
 import {is} from 'electron-util';
+import config from './config';
+import {toggleMenuBarMode} from './menu-bar-mode';
 
 let tray: Tray | null = null;
+let previousMessageCount = 0;
+
+let contextMenu: any;
```
Comment (sindresorhus) [inline source/tray.ts]: Don't use `any`.

File: source/tray.ts [hunk 7]
```diff
@@ -1,44 +1,84 @@
...
-			} else {
-				win.show();
-			}
-		};
-
-		const contextMenu = Menu.buildFromTemplate([
+		contextMenu = Menu.buildFromTemplate([
 			{
-				label: 'Toggle',
+				label: 'Disable Menu Bar Mode',
```
Comment (sindresorhus) [inline source/tray.ts]: What about Linux and Windows?

File: readme.md [hunk 8]
```diff
@@ -142,6 +143,11 @@ Note that you can change the behavior of Caprine so that the app closes when the
 
 Advanced users can modify the colors/styles of Caprine. Click the menu item `Caprine`/`File` → `Caprine Settings` → `Advanced` → `Custom Styles` and a CSS file will open up in your default editor.
 
+### Menu Bar Mode *(macOS only)*
+
+You can remove the app from the dock, create a menu bar icon and have a global shortcut to toggle the window. Caprine will show above any app, even if they are full screen, and it'll hide when is not focused.
+You can enable this in the menu bar options and disable in the options of the menu bar icon (right click on it).
```
Comment (sindresorhus) [inline readme.md]: This should better explain where to find those two options and should also mention that keyboard shortcuts will not work when the Dock icon is hidden. It should also explain where to find the menu then.

Would also be good with a screenshot of the menu bar icon.

**Comment by sindresorhus** [comment]
In the menu bar icon menu, the "Menu" item should only be visible if the Dock icon is hidden.

---

When the user unchecks the "Show Dock Icon" preference, we should show a dialog explaining to the user that keyboard shortcuts will no longer work, and that they can access the menu by right-clicking the menu bar icon.

---

Would the "Hide Dock Icon" maybe be better placed in the menu bar icon menu? Since it only really applies when the menu bar icon is shown.

**Comment by whitecrownclown** [comment]
I agree with 1&3
2) Shortcuts work even if we hide the Dock Icon since we have the Menu
registered in the Menu bar.

On Mon, 17 Jun 2019 at 14:24, Sindre Sorhus <notifications@github.com>
wrote:

> In the menu bar icon menu, the "Menu" item should only be visible if the
> Dock icon is hidden.
>
------------------------------

**Comment by sindresorhus** [comment]
> 2) Shortcuts work even if we hide the Dock Icon since we have the Menu
registered in the Menu bar.

Oh wow. You're right. I didn't think that would work. Great!

**Comment by whitecrownclown** [comment]
> If we set [electron-userland/electron-builder#1456 (comment)](https://github.com/electron-userland/electron-builder/issues/1456#issuecomment-293518692), can we then show the Dock icon? If so, we could just do the reverse of what we're doing.

I think it might work. I'll give it a shot.

File: css/dark-mode.css [hunk 9]
```diff
@@ -267,9 +267,6 @@ html.dark-mode ._1ht6 {
 }
 
 /* Contact list: person (selected) window inactive */
-html.dark-mode.is-window-inactive ._1ht1._1ht2 {
-	background: var(--base-twenty) !important;
-}
```
Comment (sindresorhus) [inline css/dark-mode.css]: ?

File: readme.md [hunk 10]
```diff
@@ -19,6 +19,7 @@ Caprine is an unofficial and privacy-focused Facebook Messenger app with many us
 - [Vibrant theme](#vibrancy)\*
 - [Privacy-focused](#privacy)
 - [Keyboard shortcuts](#keyboard-shortcuts)
+- [Menu Bar mode](#menu-bar-mode)\*
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
- [Menu bar mode](#menu-bar-mode)\*
```

File: readme.md [hunk 11]
```diff
@@ -142,6 +143,14 @@ Note that you can change the behavior of Caprine so that the app closes when the
 
 Advanced users can modify the colors/styles of Caprine. Click the menu item `Caprine`/`File` → `Caprine Settings` → `Advanced` → `Custom Styles` and a CSS file will open up in your default editor.
 
+### Menu Bar Mode *(macOS only)* <img src="media/screenshot-menu-bar-mode.png" width="40">
+
+You can create a menu bar icon and have a global shortcut to toggle the window (<kbd>Cmd</kbd> <kbd>Shift</kbd> <kbd>y</kbd>).You can also remove the app from the dock by toggling the menu item `Hide Dock Icon` from the menu bar icon.
```
Comment (sindresorhus) [inline readme.md]: You should explain where to find the setting for enabling the menu bar icon. And I think the text could be better written.
```suggestion
You can enable a menu bar icon and have a global shortcut to toggle the window (<kbd>Cmd</kbd> <kbd>Shift</kbd> <kbd>y</kbd>). You can also remove the app from the dock by toggling the menu item `Hide Dock Icon` from the menu bar icon.
```

File: readme.md [hunk 12]
```diff
@@ -142,6 +143,14 @@ Note that you can change the behavior of Caprine so that the app closes when the
...
+This will also make the app `Menu` available as a menu item inside the menu bar icon.
+
+<div align="center">
+<img src="media/screenshot-menu-bar-menu.png" width="200"></div>
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
	<img src="media/screenshot-menu-bar-menu.png" width="200">
</div>
```

File: source/index.ts [hunk 13]
```diff
@@ -289,7 +296,9 @@ function createMainWindow(): BrowserWindow {
 
 	await updateAppMenu();
 	mainWindow = createMainWindow();
-	tray.create(mainWindow);
+
+	// Start in Menu Bar mode if enabled, otherwise start normally
```
Comment (sindresorhus) [inline source/index.ts]:
```suggestion
	// Start in menu bar mode if enabled, otherwise start normally
```

**Comment by sindresorhus** [comment]
Almost there! Just some minor nitpick left :)

**Comment by sindresorhus** [comment]
Have to tested the app on Linux or Windows too? To ensure the tray functionality still works there.

**Comment by whitecrownclown** [comment]
@sindresorhus 	I did look on Windows, but can't check it on Linux. I can have another look on Windows, just to make sure. I'll also update with what you reviewed. Thanks!

File: css/dark-mode.css (hunk 9 above)
Comment (whitecrownclown) [reply css/dark-mode.css]: Hm, this is how it looks for me:
Screenshot 2019-06-20 at 22 48 41

**Comment by sindresorhus** [comment]
Yay! Great to finally have this in. Nice work, @whitecrownclown 🙌

**Pull request merged**